src/SAML2/WebSSO/Types.hs

```diff
@@ -25,7 +157,48 @@ import URI.ByteString  -- FUTUREWORK: should saml2-web-sso also use the URI from
 import qualified Data.List as L
 import qualified Data.Text as ST
 import qualified Data.X509 as X509
+import qualified Foundation.Network.IPv4 as IPv4
+import qualified Foundation.Network.IPv6 as IPv6
+import qualified Foundation.Parser as IP
+import qualified Network.DNS.Utils as DNS
 import qualified Servant
+import qualified Text.Email.Validate as Email
+
+
+-- | Text that needs to be escaped when rendered into XML.  See 'mkXmlText', 'mkXmlText'.
+newtype XmlText = XmlText { fromXmlText :: ST }
+  deriving (Eq, Ord, Show, Generic)
+
+mkXmlText :: ST -> XmlText
```

Comment:
This means that `&lt;` is not allowed anywhere in strings that we get from outside, right?

It actually seems kinda weird and I don't think it makes anything better. I would just do `mkXmlText = XmlText` and whenever an issue pops up, it's a bug that needs to be fixed. Unescaping-and-escaping just hides it.

Reply:
Yes, I got that wrong, thanks!  The xml parsers definitely should give us unescaped content, and if not, we should fix it there.

Tasks:

- [ ] test case for NameID with formats unspecified, email, entity (aka uri), with escaping corner cases.
- [ ] factor `unescapeXmlText` out of `mkXmlText`
- [ ] `mkXmlText = XmlText`, as you suggest.